feat: Enhance relpath error handling and validation

7amo10 opened the pull request:
## Changes
- Add input validation for fork numbers to prevent invalid access
- Improve memory management with proper cleanup
- Add version bound constants for better maintainability
- Optimize fork names lookup with static const array
- Add proper error codes for better error reporting

## Technical Details
- Added `is_valid_fork_number()` validation helper
- Added `MIN_PG_VERSION` and `MAX_PG_VERSION` constants
- Enhanced error reporting with `errno` codes
- Optimized string handling with static arrays
- Added memory allocation failure checks

jesperpedersen commented:
@sh-soltanpour Can you take a look ?

@7amo10 You need `./uncrustify.sh` for sure

jesperpedersen commented:
There is something up... leave out any changes that you didn't change; I mean files

sh-soltanpour commented:
yeah @7amo10 please remove unwanted changes and squash your commits based on [this guide](https://github.com/pgmoneta/pgmoneta/blob/main/CONTRIBUTING.md#development) and I'll take a look

7amo10 commented:
@sh-soltanpour 
I had to force push to remove an unwanted changes and ensure the PR stays focused on `relpath.c`. The changes remain the same functionally. Let me know if anything needs revision!

jesperpedersen commented:
@7amo10 Take a look at your patch, and remove unneeded changes. `uncrustify.sh` is also your friend

7amo10 commented:
@jesperpedersen 

I’ve updated the commit with the changes to `src/libpgmoneta/walfile/relpath.c`. I’ve kept the functional improvements (input validation, memory management, error codes, etc.) as intended, and followed it with a formatting commit using `./uncrustify.sh` per the guidelines. I hope this addresses your feedback about removing unwanted changes—if I misunderstood, please let me know.

Could you clarify if there’s anything else you’d like me to adjust or improve?

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -36,130 +36,164 @@
  */
 
-#include <walfile/relpath.h>
+ #include <walfile/relpath.h>
```
Indentation

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -36,130 +36,164 @@
+/* Constants for configuration */
+ #define MAX_VERSION_DIR_SIZE 50
+ #define MIN_PG_VERSION 13
+ #define MAX_PG_VERSION 17
```
Still extra space

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -36,130 +36,164 @@
+      return NULL;
+   }
+
+   char* path = NULL;
```
Define variables in the beginning of the function

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -36,130 +36,164 @@
+      if (dbNode != 0 || backendId != INVALID_BACKEND_ID)
       {
-         path = pgmoneta_format_and_append(path, "global/%u", relNode);
+         errno = EINVAL;
```
We don't use `errno`

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -36,130 +36,164 @@
       {
-         path = pgmoneta_format_and_append(path, "global/%u", relNode);
+         errno = EINVAL;
+         return NULL;
```
Do a `goto error;` instead

jesperpedersen commented:
Apply those comments over the entire patch

7amo10 commented:
I have applied the requested changes:

- Resolve Indentation problems and extra spaces
- Removed unnecessary modifications that were unrelated to error handling and validation.
- Ensured proper memory management by handling allocations correctly in `pgmoneta_wal_get_tablespace_version_directory()`.

Please review and let me know if any further adjustments are needed. Thanks for your guidance!

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,170 @@
 char*
 pgmoneta_wal_get_catalog_version_number(void)
 {
+   if (!server_config || server_config->version < MIN_PG_VERSION ||
```
Everything other than !server_config is already handled in the switch case

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,170 @@
+error:
+   if (result)
+   {
+      free(result);
```
if is not needed, free(NULL)
`If ptr is a null pointer, the function does nothing.`

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,170 @@
+error:
+   if (path)
+   {
+      free(path);
```
`if` is not needed

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,170 @@
       free(version_directory);
    }
+
+   if (!path)
```
why do we need the `if` here?

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,170 @@
+      goto error;
+   }
+
+   catalog_version = pgmoneta_wal_get_catalog_version_number();
```
you're already handling this in `pgmoneta_wal_get_catalog_version_number`

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,153 @@
-                                              spcNode, version_directory,
-                                              dbNode, backendId, relNode);
-         }
+         path = pgmoneta_format_and_append(path,
```
IMO the `if` and `else` approach is easier to read. WDYT? @jesperpedersen

jesperpedersen replied:
Agreed

7amo10 replied:
You both agreed that the `if` and `else` approach is easier to read. Just to confirm, are you suggesting restructuring the function to improve readability while keeping the same logic? Or do you mean that we should avoid refactoring it further since it's already structured with `if-else`? Let me know how you'd like me to proceed!

jesperpedersen replied:
Yes, keep the `if-else`

jesperpedersen replied:
Keep the `if`

jesperpedersen replied:
?

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,153 @@
+#include <stdlib.h>
+
+/* Constants for configuration */
+#define MAX_VERSION_DIR_SIZE 50
```
Move this to `relpath.h`

sh-soltanpour commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,153 @@
+/* Constants for configuration */
+#define MAX_VERSION_DIR_SIZE 50
+
+/* Validation helper for fork numbers */
```
remove unncessary comments

jesperpedersen commented:
@7amo10 Can you rebase, and fix the above ?

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,143 @@
+   return forkNumber >= MAIN_FORKNUM && forkNumber <= INIT_FORKNUM;
+}
+
+static const char* const FORK_NAMES[] = {
```
Move `{` to newline

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,143 @@
+         goto error;
       }
+
+      path = pgmoneta_format_and_append(path,
```
Do an `if`

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,143 @@
-            path = pgmoneta_format_and_append(path, "base/%u/%u",
-                                              dbNode, relNode);
-         }
+         path = pgmoneta_format_and_append(path,
```
Keep the `if`

jesperpedersen replied:
?

jesperpedersen commented on src/libpgmoneta/walfile/relpath.c:
```diff
@@ -37,129 +37,143 @@
-            path = pgmoneta_format_and_append(path, "base/%u/t%d_%u",
-                                              dbNode, backendId, relNode);
-         }
+         path = pgmoneta_format_and_append(path,
```
Keep the `if`

jesperpedersen replied:
?

jesperpedersen commented:
You need to fix the `if-else`s

jesperpedersen commented:
You need to reference the issue number in the commit subject -- see the git log for examples

7amo10 commented:
> You need to reference the issue number in the commit subject -- see the git log for examples

Please review the this and merge whenever you have time.

jesperpedersen commented:
Merged.

Thanks for your contribution !
